### PR TITLE
Opt in to trailing slash 307 redirect. Opt out of 405 logic.

### PR DESCRIPTION
**Pull request by schloerke**
Fixes https://github.com/rstudio/plumber/issues/100
Fixes https://github.com/rstudio/plumber/issues/465

If a route is submitted without a trailing slash, and a comparable route exists with a trailing slash, then a 307 redirect is returned

```r
options_plumber("redirect" = TRUE)

pr() %>% 
  pr_get("/get/", function(a) a) %>%
  pr_post("/post/", function(a) a) %>%
  pr_mount(
    "/mnt", 
    pr() %>% pr_get("/", function(a) a)
  ) %>% 
  pr_run(port = 8000)
```

```bash
$ curl "localhost:8000/mnt"
{"message":"307 - Redirecting with trailing slash"}

# follow redirect
$ curl "localhost:8000/mnt?a=42" -L
["42"]
```

```bash
curl "localhost:8000/mnt"
#> {"message":"307 - Redirecting with trailing slash"}

# follow redirect
curl "localhost:8000/mnt?a=42" -L
#> ["42"]

curl "localhost:8000/get?a=42"
#> {"message":"307 - Redirecting with trailing slash"}
curl "localhost:8000/get?a=42" -L
#> ["42"]

curl --data 'a=42' "localhost:8000/post"
#> {"message":"307 - Redirecting with trailing slash"}
#> ["42"]
```

PR task list:
- [x] Update NEWS
- [x] Add tests
- [x] Update documentation with `devtools::document()`

-------------

* I am not sold on the new option names. Any ideas?
* Is it ok to bring in `ellipsis`?
* Is stopping ok if a bad option name is given or an unnamed option is given? 
* Is ok to put `...` at the beginning of `options_plumber()`?

### CODE REVIEW

File: R/options_plumber.R [hunk 1]
```diff
@@ -34,16 +42,20 @@
...
 #' If a particular parameter is not supplied, it will return the current value.
 #' If no parameters are supplied, all returned values will be the current [options()] values.
 #' @export
 options_plumber <- function(
+  ...,
   port                 = getOption("plumber.port"),
   docs                 = getOption("plumber.docs"),
   docs.callback        = getOption("plumber.docs.callback"),
+  redirect             = getOption("plumber.redirect"),
+  allowMethods         = getOption("plumber.allowMethods"),
```
Comment (schloerke) [inline R/options_plumber.R:58]: > The method received in the request-line is known by the origin server but not supported by the target resource.


```suggestion
  trailingSlash        = getOption("plumber.trailingSlash"),
  methodNotAllowed     = getOption("plumber.methodNotAllowed"),
```

File: DESCRIPTION [hunk 2]
```diff
@@ -31,7 +31,8 @@ Imports:
     swagger (>= 3.33.0),
```
Comment (schloerke) [inline DESCRIPTION:31]: bump minor version of plumber due to `options_plumber()` api change

File: NEWS.md [hunk 3]
```diff
@@ -7,11 +7,18 @@ plumber 1.0.0.9999 Development version
 
 * When plumbing a Plumber file and using a Plumber router modifier (`#* @plumber`), an error will be thrown if the original router is not returned. (#738)
 
+* `options_plumber()` now requires that all options are named. If no option name is provided, an error with be thrown. (#746)
+
 ### New features
 
+* Added option `plumber.redirect`. This option (which is **disabled** by default) allows routes to be redirected to route definitions with a trailing slash. For example, if a `GET` request is submitted to `/test?a=1` with no `/test` route is defined, but a `GET` `/test/` route definition does exist, then the original request will respond with a `307` to reattempt against `GET` `/test/?a=1`. This option will be _enabled_ by default in a future release. This logic executed for before calling the `404` handler. (#746)
+
+* Added option `plumber.allowMethods`. This option (which is enabled by default) allows for a status of `405` to be returned if an invalid method is used when requesting a valid route. This logic executed for before calling the default `404` handler. (#746)
```
Comment (schloerke) [inline NEWS.md:16]:
```suggestion
* Added option `plumber.trailingSlash`. This option (which is **disabled** by default) allows routes to be redirected to route definitions with a trailing slash. For example, if a `GET` request is submitted to `/test?a=1` with no `/test` route is defined, but a `GET` `/test/` route definition does exist, then the original request will respond with a `307` to reattempt against `GET` `/test/?a=1`. This option will be _enabled_ by default in a future release. This logic executed for before calling the `404` handler. (#746)

* Added option `plumber.methodNotFound`. This option (which is enabled by default) allows for a status of `405` to be returned if an invalid method is used when requesting a valid route. This logic executed for before calling the default `404` handler. (#746)
```

File: R/options_plumber.R [hunk 4]
```diff
@@ -11,6 +11,14 @@
...
 #' a single parameter corresponding to the visual documentation url after Plumber server is ready. This can be used
 #' by RStudio to open the docs when then API is ran from the editor. Defaults to option `NULL`.}
+#' \item{`plumber.redirect`}{Logical value which allows the router to redirect any request
+#' that has a matching route with a trailing slash. For example, if set to `TRUE` and the
+#' GET route `/test/` existed, then a GET request of `/test?a=1` would redirect to
+#' `/test/?a=1`. Defaults to `FALSE`. This option will default to `TRUE` in a future release.}
+#' \item{`plumber.allowMethods`}{Logical value which allows the router to notify that an
+#' unavailable method was requested, but a different request method is allowed. For example,
+#' if set to `TRUE` and the GET route `/test` existed, then a POST request of `/test` would
+#' receive a 405 status and the allowed methods. Defaults to `TRUE`.}
```
Comment (schloerke) [inline R/options_plumber.R:21]:
```suggestion
#' \item{`plumber.trailingSlash`}{Logical value which allows the router to redirect any request
#' that has a matching route with a trailing slash. For example, if set to `TRUE` and the
#' GET route `/test/` existed, then a GET request of `/test?a=1` would redirect to
#' `/test/?a=1`. Defaults to `FALSE`. This option will default to `TRUE` in a future release.}
#' \item{`plumber.methodNotFound`}{Logical value which allows the router to notify that an
#' unavailable method was requested, but a different request method is allowed. For example,
#' if set to `TRUE` and the GET route `/test` existed, then a POST request of `/test` would
#' receive a 405 status and the allowed methods. Defaults to `TRUE`.}
```

File: R/options_plumber.R [hunk 5]
```diff
@@ -34,16 +42,20 @@
 #' by settings this option to `FALSE`. Defaults to `TRUE`}
 #' }
 #'
-#' @param port,docs,docs.callback,apiScheme,apiHost,apiPort,apiPath,apiURL,maxRequestSize,sharedSecret,legacyRedirects See details
+#' @param ... Ignored. Should be empty
+#' @param port,docs,docs.callback,redirect,allowMethods,apiScheme,apiHost,apiPort,apiPath,apiURL,maxRequestSize,sharedSecret,legacyRedirects See details
```
Comment (schloerke) [inline R/options_plumber.R:46]:
```suggestion
#' @param port,docs,docs.callback,trailingSlash,methodNotFound,apiScheme,apiHost,apiPort,apiPath,apiURL,maxRequestSize,sharedSecret,legacyRedirects See details
```

File: tests/testthat/test-plumber.R [hunk 6]
```diff
@@ -348,25 +348,67 @@ test_that("invalid hooks err", {
 })
 
 test_that("handle invokes correctly", {
-  pr <- pr()
-  pr$handle("GET", "/trailslash", function(){ "getter" })
-  pr$handle("POST", "/trailslashp/", function(){ "poster" })
+  with_options(
+    list(plumber.redirect = NULL),
```
Comment (schloerke) [inline tests/testthat/test-plumber.R:352]:
```suggestion
    list(plumber.trailingSlash = NULL),
```

File: tests/testthat/test-plumber.R [hunk 7]
```diff
@@ -348,25 +348,67 @@ test_that("invalid hooks err", {
...
+    expect_equal(res$status, 404)
 
+    res <- pr$call(make_req("POST", "/post", "?a=1"))
+    expect_equal(res$status, 404)
+
+    res <- pr$call(make_req("GET", "/mnt", "?a=1"))
+    expect_equal(res$status, 404)
+  })
+
+  with_options(list(plumber.redirect = TRUE), {
```
Comment (schloerke) [inline tests/testthat/test-plumber.R:396]:
```suggestion
  with_options(list(plumber.trailingSlash = FALSE), {
    res <- pr$call(make_req("GET", "/get", "?a=1"))
    expect_equal(res$status, 404)

    res <- pr$call(make_req("POST", "/post", "?a=1"))
    expect_equal(res$status, 404)

    res <- pr$call(make_req("GET", "/mnt", "?a=1"))
    expect_equal(res$status, 404)
  })

  with_options(list(plumber.trailingSlash = TRUE), {
```

**Review by cpsievert — APPROVED** [review]

File: R/options_plumber.R [hunk 8]
```diff
@@ -53,17 +65,21 @@ options_plumber <- function(
...
-    plumber.port = port,
-    plumber.sharedSecret = sharedSecret,
-    plumber.docs = docs,
-    plumber.docs.callback = docs.callback,
-    plumber.legacyRedirects = legacyRedirects
+    plumber.port                 =   port,
+    plumber.docs                 =   docs,
+    plumber.docs.callback        =   docs.callback,
+    plumber.redirect             =   redirect,
+    plumber.allowMethods         =   allowMethods,
```
Comment (schloerke) [inline R/options_plumber.R:75]:
```suggestion
    plumber.trailingSlash        =   trailingSlash,
    plumber.methodNotAllowed     =   methodNotAllowed,
```

**Pull request merged**